sql-parser/src/main/java/io/crate/sql/Literals.java

```diff
@@ -24,14 +24,191 @@
 
 import com.google.common.annotations.VisibleForTesting;
 
+import java.util.function.Predicate;
+
 public class Literals {
 
+    private static final String EPSILON = "E";
+    private static final String ESCAPED_UNICODE_ERROR = "Invalid Unicode escape (must be \\uXXXX or \\UXXXXXXXX)";
+
     public static String quoteStringLiteral(String literal) {
         return "'" + escapeStringLiteral(literal) + "'";
     }
 
+    public static String escapeAndQuoteStringLiteral(String literal) {
+        return EPSILON + Literals.quoteStringLiteral(literal);
+    }
+
     @VisibleForTesting
     static String escapeStringLiteral(String literal) {
         return literal.replace("'", "''");
     }
+
+    public static String replaceEscapedChars(String input) {
+        if (input.length() <= 1) {
+            return input;
+        }
+
+        char[] inputCharArray = input.toCharArray();
+        StringBuilder builder = new StringBuilder(inputCharArray.length);
+        int current = 0, prev = 0;
+
+        boolean firstToken = true;
+        int nextTokenIdx, tokenBeginIdx, tokenLen;
+
+        while (current < inputCharArray.length) {
```

Comment:
I'm probably missing something here, but wouldn't it be possible to go through the characters one by one like this:

```
        StringBuilder result = new StringBuilder();
        int length = input.length();
        for (int i = 0; i < length; i++) {
            char currentChar = input.charAt(i);
            if (currentChar == '\\' && i + 1 < length) {
                char nextChar = input.charAt(i + 1);
                case (nextChar) {
                ....
```

And if there is a escaped sequence do something along the lines of

```
                    int codePoint = Integer.parseInt(x, i + 2, i + 2 + numCharsInSeq, 16);
                    result.append(Character.toChars(codePoint));
                    i += 1 + numCharsInSeq; // skip the extra chars consumed
```

This avoids most copies, doesn't require a custom `parseInt` and should be fairly straight forward?

Reply:
thanks for the input. It's true, no reason to further keep the tokenization logic, what you proposed makes much more sense (and we do get rid of the custom `parseInt` !). +1